### PR TITLE
Fix formatting of decimal numbers in BF43COMPAT display mode

### LINKED ISSUE #8769: BF43COMPAT Current broken above 10A
With BF43COMPAT current Current OSD element is broken above 10A. See bottom left corner

### PR DESCRIPTION
**Pull request by mmosca**
Add unit test to osdFormatCentiNumber() to verify changes.

Handle corner case where there would be no space left for any decimals so generated string is the correct length.

I suspect the corruption in the osd described in #8769 has been caused by the last digit of the previous screen not being overwritten.

The bulk of the changes are re-arranging things to enable running the unit test.

### CODE REVIEW

File: src/main/io/osd.c [hunk 1]
```diff
@@ -33,6 +33,31 @@
...
 FILE_COMPILE_FOR_SPEED
 
+#include "drivers/osd_symbols.h"
+#include "common/maths.h"
+#include "io/displayport_msp_bf_compat.h"
+#include "config/parameter_group.h"
+#include "config/parameter_group_ids.h"
+#include "common/typeconversion.h"
+
+//PG_REGISTER_WITH_RESET_TEMPLATE(osdConfig_t, osdConfig, PG_OSD_CONFIG, 7);
```
Comment (MrD-RC) [inline src/main/io/osd.c]: PG register is lower down in the file. These commented out lines should be removed as they are duplicates and add confusion. It would be good if the PG lines were just after the includes/defines, before the function blocks.

**Review by MrD-RC — CHANGES_REQUESTED** [review]

**Comment by mmosca** [comment]
I will do that. I will probably move the functions needed by the unit test to a separate file to simplify the ifdefs and clean up the changes in osd.c[]s,Marcelo Bezerra ***@***.***>On 8 Feb 2023, at 07:31, Darren Lines ***@***.***> wrote:﻿
@MrD-RC requested changes on this pull request.



In src/main/io/osd.c:
> @@ -33,6 +33,31 @@
 
 FILE_COMPILE_FOR_SPEED
 
+#include "drivers/osd_symbols.h"
+#include "common/maths.h"
+#include "io/displayport_msp_bf_compat.h"
+#include "config/parameter_group.h"
+#include "config/parameter_group_ids.h"
+#include "common/typeconversion.h"
+
+//PG_REGISTER_WITH_RESET_TEMPLATE(osdConfig_t, osdConfig, PG_OSD_CONFIG, 7);

PG register is lower down in the file. These commented out lines should be removed as they are duplicates and add confusion. It would be good if the PG lines were just after the includes/defines, before the function blocks.

**Comment by mmosca** [comment]
> @mmosca ready to merge? I'd love to build RC2 tomorrow

The code is ready but I haven't tested the code on actual hardware yet.

File: src/main/io/osd_utils.c [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    // Done counting. Time to write the characters.
+    // Write spaces at the start
+    while (remaining > 0) {
+        *ptr = SYM_BLANK;
+        ptr++;
+        remaining--;
+    }
+
+    // Keep number right aligned and correct length
```
Comment (mmosca) [inline src/main/io/osd_utils.c:96]: This is the only actual change in the formatting function

**Review by MrD-RC — APPROVED** [review]
